Source/WTF/wtf/NativePromise.h

```diff
@@ -1466,23 +1466,37 @@ class NativePromiseRequest final {
     RefPtr<typename PromiseType::Request> m_request;
 };
 
+template<typename S, typename E>
+Ref<NativePromise<S, E>> createSettledPromise(Expected<S, E>&& result)
+{
+    return NativePromise<S, E>::createAndSettle(WTFMove(result));
+}
+
 // Invoke a function object (e.g., lambda) asynchronously.
 // Returns a promise that the function should eventually resolve or reject once the original promise returned by the lambda
 // is itself resolved or rejected.
 template<typename Function>
 static auto invokeAsync(SerialFunctionDispatcher& targetQueue, Function&& function, const Logger::LogSiteIdentifier& callerName = DEFAULT_LOGSITEIDENTIFIER)
 {
     static_assert(!std::is_lvalue_reference_v<Function>, "Function object must not be passed by lvalue-ref (to avoid unplanned copies); WTFMove() the object.");
-    static_assert(IsConvertibleToNativePromise<typename RemoveSmartPointer<decltype(function())>::type>, "Function object must return Ref<NativePromise>");
     using ReturnType = typename RemoveSmartPointer<decltype(function())>::type;
+    static_assert(IsConvertibleToNativePromise<ReturnType> || IsExpected<ReturnType>::value, "Function object must return Ref<NativePromise> or Expected<T, F>");
 
-    typename ReturnType::PromiseType::Producer proxyPromiseProducer(PromiseDispatchMode::Default, callerName);
-    auto promise = proxyPromiseProducer.promise();
-    targetQueue.dispatch([producer = WTFMove(proxyPromiseProducer), function = WTFMove(function)] () mutable {
-        Ref<typename ReturnType::PromiseType> p = function();
-        p->chainTo(WTFMove(producer), { "invokeAsync proxy", nullptr });
-    });
-    return promise;
+    if constexpr (IsConvertibleToNativePromise<ReturnType>) {
+        typename ReturnType::PromiseType::Producer proxyPromiseProducer(PromiseDispatchMode::Default, callerName);
+        auto promise = proxyPromiseProducer.promise();
+        targetQueue.dispatch([producer = WTFMove(proxyPromiseProducer), function = WTFMove(function)] () mutable {
+            static_cast<Ref<typename ReturnType::PromiseType>>(function())->chainTo(WTFMove(producer), { "invokeAsync proxy", nullptr });
+        });
+        return promise;
+    } else {
```

Comment:
No need for else.
Would be good to add a code path for `void` as well to remove an unnecessary `GenericPromise::createAndResolve()` call.

Reply:
ok will do.